src/org/jgroups/raft/blocks/LockService.java

```diff
@@ -0,0 +1,845 @@
+	 * @param lockId the lock's id
+	 * @return lock status
+	 */
+	public LockStatus lockStatus(long lockId) {
+		LockStatus v = lockStatus.get(lockId); return v == null ? NONE : v;
+	}
+
+	/**
+	 * Acquire the lock, will join the waiting queue if the lock is held by another member currently.
+	 * @param lockId the lock's id
+	 * @return HOLDING if hold the lock, WAITING if in the waiting queue.
+	 * @throws RaftException if exception happens during sending or executing commands in the lock service.
+	 */
+	public CompletableFuture<LockStatus> lock(long lockId) {
+		var out = new ByteArrayDataOutputStream(26);
+		out.writeByte(LOCK);
+		writeLong(lockId, out);
+		writeUuid(address(), out);
+		assert out.position() <= 26;
+		return invoke(out).thenApply(t -> notifyListeners(lockId, LockStatus.values()[t[0]]));
+	}
+
+	/**
+	 * Try to acquire the lock, won't join the waiting queue.
+	 * @param lockId the lock's id
+	 * @return HOLDING if hold the lock, NONE if the lock is held by another member.
+	 * @throws RaftException if exception happens during sending or executing commands in the lock service.
+	 */
+	public CompletableFuture<LockStatus> tryLock(long lockId) {
+		var out = new ByteArrayDataOutputStream(26);
+		out.writeByte(TRY_LOCK);
+		writeLong(lockId, out);
+		writeUuid(address(), out);
+		assert out.position() <= 26;
+		return invoke(out).thenApply(t -> notifyListeners(lockId, LockStatus.values()[t[0]]));
+	}
+
+	/**
+	 * Release the lock if it's the holder, and take next waiting member from the queue to be the new holder if there
+	 * is one. Remove from waiting queue if it's waiting. Do nothing if neither of them.
+	 * @param lockId the lock's id
+	 * @return async completion
+	 * @throws RaftException if exception happens during sending or executing commands in the lock service.
+	 */
+	public CompletableFuture<Void> unlock(long lockId) {
+		var out = new ByteArrayDataOutputStream(26);
+		out.writeByte(UNLOCK);
+		writeLong(lockId, out);
+		writeUuid(address(), out);
+		assert out.position() <= 26;
+		return invoke(out).thenApply(t -> { notifyListeners(lockId, NONE); return null; });
+	}
+
+	/**
+	 * Release all related locks for this member.
+	 * @return async completion
+	 * @throws RaftException if exception happens during sending or executing commands in the lock service.
+	 */
+	public CompletableFuture<Void> unlockAll() {
+		var out = new ByteArrayDataOutputStream(17);
+		out.writeByte(UNLOCK_ALL);
+		writeUuid(address(), out);
+		assert out.position() <= 17;
+		return invoke(out).thenApply(t -> { resign(); return null; });
+	}
+
+	protected UUID address() {
+		return Objects.requireNonNull(address);
+	}
+
+	protected CompletableFuture<byte[]> invoke(ByteArrayDataOutputStream out) {
+		try {
+			return raft.setAsync(out.buffer(), 0, out.position());
+		} catch (Throwable e) {
+			throw new RaftException("Fail to execute command", e);
+		}
+	}
+
+	protected LockStatus notifyListeners(long lockId, LockStatus curr) {
```

Comment:
In leader, the calling thread will be the RAFT working thread. In follower it's the thread that pass up the message that send from leader, I think they are both ordered and synchronized even stable because that's what JGroups do.